subsystems/motor/speedmodifiers/AccelerationCap.java

```diff
@@ -41,15 +52,69 @@ public AccelerationCap(PDP pdp, double softStopVoltage, double hardStopVoltage)
 	 * This is designed to reduce power consumption (via voltage ramping)
 	 * and prevent RoboRIO/router brownouts.
 	 *
-	 * Default soft stop voltage is 11 volts.
-	 * Default hard stop voltage is 10 volts.
+	 * Default hard stop voltage is 9.0 volts.
 	 *
 	 * @param pdp
 	 *        The robot's power distribution panel.
 	 *        This is used to monitor the battery voltage.
 	 */
-	public AccelerationCap(PDP pdp) {
-		this(pdp, 11.0, 10.0);
+	public AccelerationCap(PDP pdp, int... channels) {
+		this(pdp, AccelerationCap.DEFAULT_HARD_STOP_VOLTAGE, channels);
+	}
+
+	protected double calculate(double inputSpeed) {
+		double deltaTime = (System.currentTimeMillis() - lastUpdate) / 1000.0;
+		lastUpdate = System.currentTimeMillis();
+		if (deltaTime > AccelerationCap.TIMEOUT_SECONDS) {
+			return 0;
+		}
+		if (Math.abs(inputSpeed) < Math.abs(currentSpeed) && Math.signum(inputSpeed) == Math.signum(currentSpeed)) {
+			// Ramp down (faster) for the sake of the gearboxes
+			if (Math.abs(currentSpeed - inputSpeed) < AccelerationCap.MAXIMUM_MOTOR_DECREASE_PER_SECOND * deltaTime) {
+				return inputSpeed;
+			}
+			if (inputSpeed > currentSpeed) {
+				return currentSpeed + AccelerationCap.MAXIMUM_MOTOR_DECREASE_PER_SECOND * deltaTime;
+			}
+			if (inputSpeed < currentSpeed) {
+				return currentSpeed - AccelerationCap.MAXIMUM_MOTOR_DECREASE_PER_SECOND * deltaTime;
+			}
+		}
+		double rampedSpeed = inputSpeed;
+		// Ramping
+		if (Math.abs(currentSpeed - inputSpeed) > AccelerationCap.MAXIMUM_MOTOR_INCREASE_PER_SECOND * deltaTime) {
+			if (inputSpeed > currentSpeed) {
+				rampedSpeed = currentSpeed + AccelerationCap.MAXIMUM_MOTOR_INCREASE_PER_SECOND * deltaTime;
+			} else if (inputSpeed < currentSpeed) {
+				rampedSpeed = currentSpeed - AccelerationCap.MAXIMUM_MOTOR_INCREASE_PER_SECOND * deltaTime;
+			}
+		}
+		// After ramping, apply brown-out protection
+		double currentVoltage = pdp.getVoltage(); // Allow fallback to DS voltage
+		if (currentVoltage < hardStopVoltage) { // If we are below hardStopVoltage, start backing off
+			rampedSpeed = currentSpeed
+				- AccelerationCap.ANTI_BROWNOUT_BACKOFF_PER_SECOND * Math.signum(currentSpeed) * deltaTime;
+			if (Math.abs(rampedSpeed) <= AccelerationCap.ANTI_BROWNOUT_BACKOFF_PER_SECOND) {
+				return 0;
+			}
+			return rampedSpeed;
+		}
+		// Even if we are still above the hard stop voltage, try to avoid going below next tick
+		try {
+			double currentCurrent = 0.0;
+			for (int channel : channels) {
+				currentCurrent += pdp.getCurrent(channel);
+			}
+			double voltageDrop = currentCurrent * pdp.getBatteryResistanceSafely();
+			double deltaVoltageDrop = voltageDrop - lastVoltageDrop;
+			if (currentVoltage < hardStopVoltage + voltageDrop + deltaVoltageDrop * AccelerationCap.TICKS_PER_PDP_DATA) {
+				return currentSpeed;
```

Comment:
So we maintain current speed if just above hardStopVoltage but ramp down if below?